### PR TITLE
TEL-4 Telemetry Agent initial implementation

### PR DESCRIPTION
**Pull request by maxkondr**
Percona Telemetry Agent initial implementation.

https://perconadev.atlassian.net/browse/TEL-4

Depends on: https://github.com/percona/saas/pull/129

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package config provides functionality for processing Telemetry Agent configuration parameters.
+package config
+
+import (
+	"path/filepath"
+
+	"github.com/spf13/viper"
```
Comment (artemgavrilov) [inline config/config.go:22]: Some time ago we choose  https://github.com/alecthomas/kong for CLI/Flags handling in Portal/PMM/Exporters. Is it possible to use it here as well?

File: cmd/telemetry-agent/main.go [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+	l.Info("scraping host metrics")
+	hostMetrics, err := metrics.ScrapeHostMetrics()
+	if err != nil {
+		l.Errorw("failed to scrape host metrics", zap.Error(err))
+		return err
+	}
+	hostInstanceID := hostMetrics.Metrics["instanceId"]
+	// instanceId is not needed in main metrics set
+	delete(hostMetrics.Metrics, "instanceId")
```
Comment (artemgavrilov) [inline cmd/telemetry-agent/main.go:123]: `instanceId` candidate for const

File: cmd/telemetry-agent/main.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+		l.DPanic(err)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	l.Info("Percona Telemetry Agent started")
+	var wg sync.WaitGroup
+	wg.Add(1)
+	utils.SignalRunner(
+		func() {
```
Comment (artemgavrilov) [inline cmd/telemetry-agent/main.go:242]: For me it looks upside down. In most our projects signal handling happens in separate goroutine and main application logic stays straightforward.

File: cmd/telemetry-agent/main.go [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+			checkIntv := time.Duration(conf.TelemetryCheckInterval) * time.Second
```
Comment (artemgavrilov) [inline cmd/telemetry-agent/main.go:243]:
```suggestion
			checkIntv := conf.TelemetryCheckInterval * time.Second
```

File: cmd/telemetry-agent/main.go [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+			l.Info(fmt.Sprintf("sleeping for %d seconds before first iteration", checkIntv))
```
Comment (artemgavrilov) [inline cmd/telemetry-agent/main.go:244]:
```suggestion
			l.Infof("sleeping for %d seconds before first iteration", checkIntv)
```

File: config/config.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+// InitConfig parses Telemetry Agent configuration parameters.
+// If some parameters are not defined - default values are used instead.
+func InitConfig() Config {
+	viper.SetEnvPrefix(perconaTelemetryEnvPrefix)
+	var err error
+
+	pillarMetricsRootPathVarName := "root_path"
+
+	pillarMetricsRootPathDefault := filepath.Join("/usr", "local", "percona", "telemetry")
+	err = viper.BindEnv(pillarMetricsRootPathVarName)
```
Comment (artemgavrilov) [inline config/config.go:55]: Can be `MustBindEnv` instead

File: config/config.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	PerconaTelemetryURL          string
+}
+
+// InitConfig parses Telemetry Agent configuration parameters.
+// If some parameters are not defined - default values are used instead.
+func InitConfig() Config {
+	viper.SetEnvPrefix(perconaTelemetryEnvPrefix)
+	var err error
+
+	pillarMetricsRootPathVarName := "root_path"
```
Comment (artemgavrilov) [inline config/config.go:52]: Better to have them as const and in uppercase, to simplify search in codebase

File: metrics/history.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+func CleanupMetricsHistory(historyPath string, keepInterval int) error {
+	l := zap.L().Sugar()
+
+	cleanHistoryPath := filepath.Clean(historyPath)
+	files, err := os.ReadDir(cleanHistoryPath)
+	if err != nil {
+		return errors.Wrap(err, "can't read directory with history metric files")
+	}
+
+	timeThreshold := time.Now().Add(-time.Duration(keepInterval) * time.Second)
```
Comment (artemgavrilov) [inline metrics/history.go:72]:
```suggestion
	timeThreshold := time.Now().Add(-(keepInterval * time.Second))
```

File: utils/signal_runner.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+// SignalRunner runs a runner function until an interrupt signal is received, at which point it
+// will call stopper.
+func SignalRunner(runner, stopper func()) {
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		runner()
+	}()
+
```
Comment (artemgavrilov) [inline utils/signal_runner.go:36]:
```suggestion
	go runner()

```

File: .gitignore [hunk 10]
```diff
@@ -0,0 +1,81 @@
+### VisualStudioCode template
```
[the remark this answers is not in the document]
Comment (maxkondr) [reply .gitignore:1]: generated by Jetbrains Golang IDE

File: .golangci.yml [hunk 11]
```diff
@@ -0,0 +1,47 @@
+---
+# Almost all linters; some of them are optional.
+
+linters-settings:
+  depguard:
+    rules:
+      main:
+        deny:
+          - pkg: errors
+            desc: Use "github.com/pkg/errors" instead
```
[the remark this answers is not in the document]
Comment (maxkondr) [reply .golangci.yml:10]: done

File: cmd/telemetry-agent/main.go [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package main is the entry point of the service
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	url2 "net/url"
```
[the remark this answers is not in the document]
Comment (maxkondr) [reply cmd/telemetry-agent/main.go:23]: done

File: cmd/telemetry-agent/main.go [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	platformReporter "github.com/percona-platform/platform/gen/telemetry/generic"
+	platformClient "github.com/percona-platform/platform/pkg/client"
+	platformLogger "github.com/percona-platform/platform/pkg/logger"
```
[the remark this answers is not in the document]
Comment (maxkondr) [reply cmd/telemetry-agent/main.go:32]: replaced with `saas`

File: cmd/telemetry-agent/main.go [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+	}
+	if u.Scheme == "" || u.Host == "" {
+		return nil, errors.New("Percona Platform Telemetry URL is invalid: scheme or host is missed")
+	}
+
+	return platformClient.New(
+		platformClient.WithLogger(zap.L().Named("perconaPlatformClient").Sugar()),
+		platformClient.WithBaseURL(u.Scheme+"://"+u.Host),
+		platformClient.WithLogFullRequest(),
+		platformClient.WithResendTimeout(time.Second*time.Duration(c.TelemetryResendTimeout)),
```
[the remark this answers is not in the document]
Comment (maxkondr) [reply cmd/telemetry-agent/main.go:72]: no,
```
./main.go:73:36: invalid operation: time.Second * c.TelemetryResendTimeout (mismatched types time.Duration and int)
````

File: metrics/pillar.go [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+package metrics
+
+import (
+	platformReporter "github.com/percona-platform/platform/gen/telemetry/generic"
+)
+
+// ProcessPSMetrics processes PS metrics and returns slice of *File.
+// Each File corresponds to a separate metrics file.
+func ProcessPSMetrics(path string) ([]*File, error) {
+	return processMetricsDirectory(path, platformReporter.ProductFamily_PRODUCT_FAMILY_PS)
```
[the remark this answers is not in the document]
Comment (maxkondr) [reply metrics/pillar.go:25]: didn't get the problem, it make code unified as soon as the same logic shall be applied several times

File: metrics/metrics.go [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+	Filename      string
+	Timestamp     time.Time
+	ProductFamily platformReporter.ProductFamily
+	Metrics       map[string]string
+}
+
+func processMetricsDirectory(path string, productFamily platformReporter.ProductFamily) ([]*File, error) {
+	l := zap.L().Sugar()
+
+	toReturn := make([]*File, 0, 5)
```
[the remark this answers is not in the document]
Comment (maxkondr) [reply metrics/metrics.go:44]: replaced with 1

File: metrics/metrics.go [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+			zap.Error(err))
+		return nil, errors.Wrap(err, "can't read directory with metric files")
+	}
+
+	if len(files) == 0 {
+		l.Infow("pillar metric directory is empty, skipping", zap.String("directory", cleanMetricsDirectoryPath))
+		return toReturn, nil
+	}
+
+	for _, file := range files {
```
[the remark this answers is not in the document]
Comment (maxkondr) [reply metrics/metrics.go:64]: done

File: metrics/metrics.go [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+	if err != nil {
+		fl.Errorw("error during opening metrics file", zap.Error(err))
+		return nil, err
+	}
+	defer func(file *os.File, fl *zap.SugaredLogger) {
+		err := file.Close()
+		if err != nil {
+			fl.Errorw("error during closing metrics file", zap.Error(err))
+		}
+	}(file, fl)
```
[the remark this answers is not in the document]
Comment (maxkondr) [reply metrics/metrics.go:101]: done

File: metrics/metrics.go [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+
+	file, err := os.Open(cleanPath)
+	if err != nil {
+		fl.Errorw("error during opening metrics file", zap.Error(err))
+		return nil, err
+	}
+	defer func(file *os.File, fl *zap.SugaredLogger) {
+		err := file.Close()
+		if err != nil {
+			fl.Errorw("error during closing metrics file", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (maxkondr) [reply metrics/metrics.go:99]: done

File: metrics/metrics.go [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+		fileMetrics.ProductFamily = productFamily
+		toReturn = append(toReturn, fileMetrics)
+	}
+
+	return toReturn, nil
+}
+
+func parseMetricsFile(path string) (*File, error) {
+	cleanPath := filepath.Clean(path)
+	fl := zap.L().Sugar().With(zap.String("file", cleanPath))
```
[the remark this answers is not in the document]
Comment (maxkondr) [reply metrics/metrics.go:89]: done

File: config/config.go (hunk 1 above)
Comment (maxkondr) [reply config/config.go:22]: will check in next commits

File: cmd/telemetry-agent/main.go (hunk 2 above)
Comment (maxkondr) [reply cmd/telemetry-agent/main.go:123]: done

File: cmd/telemetry-agent/main.go (hunk 4 above)
Comment (maxkondr) [reply cmd/telemetry-agent/main.go:243]: no, see above

File: cmd/telemetry-agent/main.go (hunk 5 above)
Comment (maxkondr) [reply cmd/telemetry-agent/main.go]: done

File: config/config.go (hunk 7 above)
Comment (maxkondr) [reply config/config.go:52]: done

File: metrics/history.go (hunk 8 above)
Comment (maxkondr) [reply metrics/history.go:82]: no, see above

File: utils/signal_runner.go (hunk 9 above)
Comment (maxkondr) [reply utils/signal_runner.go:36]: done

File: config/config.go (hunk 6 above)
Comment (maxkondr) [reply config/config.go]: done

File: .gitignore (hunk 10 above)
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply .gitignore:1]: Up to you, but I prefer to have only necessary items in `.gitignore` (and everywhere else)

**Review by artemgavrilov — APPROVED** [review]

**Review by artemgavrilov — APPROVED** [review]

**Pull request merged**